src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,658 @@
+        }
+        throw new NoSuchAlgorithmException(
+                "Algorithm " + algorithm + " not available");
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm from
+     * the specified provider and is initialized with the specified parameters.
+     * The specified provider must be registered in the security provider list.
+     *
+     * @param algorithm
+     *         the key derivation algorithm to use. See the {@code KDF} section
+     *         in the <a href="{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *         Java Security Standard Algorithm Names Specification</a> for
+     *         information about standard KDF algorithm names.
+     * @param kdfParameters
+     *         the {@code KDFParameters} used to configure the derivation
+     *         algorithm or {@code null} if no parameters are provided
+     * @param provider
+     *         the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *         if the specified provider does not support the specified
+     *         {@code KDF} algorithm
+     * @throws NoSuchProviderException
+     *         if the specified provider is not registered in the security
+     *         provider list
+     * @throws InvalidAlgorithmParameterException
+     *         if the specified provider does not support a {@code KDF}
+     *         implementation for the specified algorithm and parameters
+     * @throws NullPointerException
+     *         if {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters,
+                                  String provider)
+            throws NoSuchAlgorithmException, NoSuchProviderException,
+                   InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "algorithm must not be null");
+        Objects.requireNonNull(provider, "provider must not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new NoSuchProviderException(msg);
+            }
+            return new KDF(new Delegate((KDFSpi) instance.impl,
+                                        instance.provider), algorithm,
+                           kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
```

Comment:
Is there really scenario requiring `handleException()`? I mean, is there really `NoSuchAlgorithmException` object containing `InvalidAlgorithmParameterException` as cause and need to be thrown as  `InvalidAlgorithmParameterException`? Can you show a possible call path for this to happen?